google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -1037,4 +1036,49 @@ private IOException asIoException(StorageException oops) {
     }
     return new IOException(oops.getMessage(), oops);
   }
+
+  /** Resolve the version of google-cloud-nio for inclusion in request meta-data */
+  private static String getVersion() {
+    // attempt to read the library's version from a properties file generated during the build
+    // this value should be read and cached for later use
+    String version = "";
+    try (InputStream inputStream =
+        CloudStorageFileSystemProvider.class.getResourceAsStream(
+            "/META-INF/maven/com.google.cloud/google-cloud-nio/pom.properties")) {
+      if (inputStream != null) {
+        final Properties properties = new Properties();
+        properties.load(inputStream);
+        version = properties.getProperty("version");
+      }
+    } catch (IOException e) {
+      // ignore
+    }
+    return version;
+  }
+
+  @VisibleForTesting
+  void doInitStorage() {
+    final StorageOptions so =
+        storageOptions != null ? storageOptions : StorageOptions.getDefaultInstance();
+    final StorageOptions.Builder builder = so.toBuilder();
+    this.storage = builder.setHeaderProvider(getHeaderProvider()).build().getService();
+  }
+
+  @VisibleForTesting
+  NioUserAgentEntryHeaderProvider getHeaderProvider() {
+    return new NioUserAgentEntryHeaderProvider();
+  }
+
+  @VisibleForTesting
+  private static class NioUserAgentEntryHeaderProvider extends FixedHeaderProvider {
+
+    private final ImmutableMap<String, String> userAgent =
+        ImmutableMap.of("user-agent", String.format("gcloud-java-nio/%s", VERSION));
```

Comment:
Does this prepend to the existing user-agent?

Reply:
It does. This is what it looks like in the logs (the version is missing because I ran it from my ide without the version file artifact
```
User-Agent: gcloud-java-nio/0.123.18-SNAPSHOT gcloud-java/2.2.1 Google-API-Java-Client/1.32.2 Google-HTTP-Java-Client/1.40.1 (gzip)
```